openstack/orchestration/v1/resourcetypes/results.go

```diff
@@ -0,0 +1,163 @@
+
+// AttributeSchema is the schema of a resource attribute
+type AttributeSchema struct {
+	Description string       `json:"description,omitempty"`
+	Type        PropertyType `json:"type"`
+}
+
+// MinMaxConstraint is a type of constraint with minimum and maximum values.
+// This is used for both Range and Length constraints.
+type MinMaxConstraint struct {
+	Min float64 `json:"min,omitempty"`
+	Max float64 `json:"max,omitempty"`
+}
+
+// ModuloConstraint constrains an integer to have a certain value given a
+// particular modulus.
+type ModuloConstraint struct {
+	Step   int `json:"step,omitempty"`
+	Offset int `json:"offset,omitempty"`
+}
+
+// ConstraintSchema describes all possible types of constraints. Besides the
+// description, only one other field is ever set at a time.
+type ConstraintSchema struct {
+	Description      string            `json:"description,omitempty"`
+	Range            *MinMaxConstraint `json:"range,omitempty"`
+	Length           *MinMaxConstraint `json:"length,omitempty"`
+	Modulo           *ModuloConstraint `json:"modulo,omitempty"`
+	AllowedValues    *[]interface{}    `json:"allowed_values,omitempty"`
+	AllowedPattern   *string           `json:"allowed_pattern,omitempty"`
+	CustomConstraint *string           `json:"custom_constraint,omitempty"`
+}
+
+// PropertySchema is the schema of a resource property.
+type PropertySchema struct {
+	Type          PropertyType              `json:"type"`
+	Description   string                    `json:"description,omitempty"`
+	Default       interface{}               `json:"default,omitempty"`
+	Constraints   []ConstraintSchema        `json:"constraints,omitempty"`
+	Required      bool                      `json:"required"`
+	Immutable     bool                      `json:"immutable"`
+	UpdateAllowed bool                      `json:"update_allowed"`
+	Schema        map[string]PropertySchema `json:"schema,omitempty"`
+}
+
+// SupportStatusDetails contains information about the support status of the
+// resource and its history.
+type SupportStatusDetails struct {
+	Status         SupportStatus         `json:"status"`
+	Message        string                `json:"message,omitempty"`
+	Version        string                `json:"version,omitempty"`
+	PreviousStatus *SupportStatusDetails `json:"previous_status,omitempty"`
+}
+
+// ResourceSchema is the schema for a resource type, its attributes, and
+// properties.
+type ResourceSchema struct {
+	ResourceType  string                     `json:"resource_type"`
+	SupportStatus SupportStatusDetails       `json:"support_status"`
+	Attributes    map[string]AttributeSchema `json:"attributes"`
+	Properties    map[string]PropertySchema  `json:"properties"`
+}
+
+// ListResult represents the result of a List operation.
+type ListResult struct {
+	gophercloud.Result
+	hasDescriptions bool
+}
+
+// Extract returns a ResourceTypeList object and is called after a
+// List operation.
+func (r ListResult) Extract() (rtl ResourceTypeList, err error) {
```

Comment:
We should be able to do the same thing like this (also taking into account the comment about modeling the response):

```go
func (r ListResult) Extract() ([]ResourceTypeResult, error) {
  var s1 struct {
    ResourceTypes []ResourceTypeResult `json:"resource_types"`
  }

  err := r.ExtractInto(&s1)
  if err == nil {
    return s1.ResourceTypes, nil
  }

  var s2 struct {
    ResourceTypes []string `json:"resource_types"`
  }

  err = r.ExtractInto(&s2)
  if err != nil {
    return nil, err
  }

  var rtr []ResourceTypeResult
  for _, i := range s2.ResourceTypes {
    v := ResourceTypeResult{
      ResourceType: i,
    }
    rtr = append(rtr, v)
  }

  return rtr, err
}
```